subsys/bluetooth/host/buf.c

```diff
@@ -85,7 +85,6 @@ struct net_buf *bt_buf_get_cmd_complete(k_timeout_t timeout)
 
 	key = irq_lock();
 	buf = bt_dev.sent_cmd;
-	bt_dev.sent_cmd = NULL;
```

Comment:
Why is this needed?

Reply:
the cmd will be clear only when corresponding response received, so when we received unexpected response(host set number completed packet response) we can also continue to receive response.

Reply:
@LingaoM It looks like `send_cmd` may also clear `bt_dev.sent_cmd` and set it to a new command. Is there any way that we will clear the wrong `bt_dev.sent_cmd` here?

Reply:
- host send host number of complete package command.

- host send scan param set command

- Controller response command complete for host number of complete command.

- Controller response command complete for scan param set command.

Reply:
send_cmd will be clear by received `Controller response command complete for host number of complete command.`, but this not we expect to do.

Reply:
> Controller response command complete for host number of complete command.

The controller won't normally send this, unless there is an error, right? 

I am not sure what the potential side effects of moving this is, or what it fixes, so I vary about approving it.

Reply:
@Thalley @rugeGerritsen

Reply:
Sorry for the late response @LingaoM  but I've been trying to really follow your comment and how this variable is used, when it is cleared and when it is set, and I still haven't been able to wrap my head around it (even after looking at this over several days). Since I still don't completely follow the change, I can't approve of this just yet, but I won't block it if e.g. @rugeGerritsen and @joerchan  can sign off on it.

Reply:
@Thalley I need to explain the reasons in detail:

We have two types of Host send command. One is to send immediately, the other is to send sequentially.

For _BT_HCI_OP_HOST_NUM_COMPLETED_PACKETS_ directly call **bt_send** to send (refer to **bt_hci_cmd_send** https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/hci_core.c#L265).

For other opcodes, they are put into **cmd_tx_queue**, and then processed in the task of **hci_tx_thread::send_cmd**.

The **sent cmd** will be recorded through **bt_dev.send_cmd** (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/hci_core.c#L2342).
Which is used to locate the corresponding command when command complete or command status is received in (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/buf.c#L97).

At that time, we made a mistake here that we took for granted.
We believe that the cmd response we received at this time was generated by the command we just sent(https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/buf.c#L97). 

As I said in the previous commit, this is not exactly the case:

As shown in the screenshot of issure #38054 :

After Host send _host number complete packets_, _LE Set Scan Params_ was sent **immediately** and at this time, our `bt.dev.send_cmd=buf(LE Set Scan Params)`.

Then we received _HCI Command Complete ( Invalid HCI Command params for BT_HCI_OP_HOST_NUM_COMPLETED_PACKETS)_, 
this is not what we expect to receive, so it should be discarded(https://github.com/zephyrproject-rtos/zephyr/pull/38056/files#diff-774b3de86adb275da54becf2a0958a2d4dae03d63e47840037ecc575e9d658b4R2011).

I think we should continue to be able to receive the Command complete corresponding to _LE Set Scan Params_, but if we are receiving HCI After Command Complete (Invalid HCI Command params for BT_HCI_OP_HOST_NUM_COMPLETED_PACKETS), set **bt_dev.send_cmd=NULL**(https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/buf.c#L98), 

Then the subsequent HCI Command complete will not be processed correctly (because this response cannot find the corresponding Command first(https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/hci_core.c#L2003).

Which eventually leads to **k_sem_take failed with err** in https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/hci_core.c#L304).

Reply:
@carlescufi @joerchan Please also take a look.

Reply:
Thanks @LingaoM I think I get it now :)